Mathlib/Analysis/BoxIntegral/UnitPartition.lean

```diff
@@ -0,0 +1,286 @@
+
+* `BoxIntegral.unitPartition.prepartition_isPartition`: For `B : BoxIntegral.Box`, if `B`
+has integral corners, the prepartition of `unitPartition.Box` admissible for `B` is a partition
+of `B`.
+-/
+
+noncomputable section
+
+variable {ι : Type*}
+
+section hasIntegralCorners
+
+open Bornology
+
+/-- A `BoxIntegral.Box` has integral devices if its corners has coordinates in `ℤ`. -/
+def BoxIntegral.hasIntegralCorners (B : Box ι) : Prop :=
+  ∃ l u : ι → ℤ, (∀ i, B.lower i = l i) ∧ (∀ i, B.upper i = u i)
+
+/-- Any bounded set is contained in a `BoxIntegral.Box` with integral corners. -/
+theorem BoxIntegral.le_hasIntegralVertices_of_isBounded [Finite ι] {s : Set (ι → ℝ)}
+    (h : IsBounded s) :
+    ∃ B : BoxIntegral.Box ι, hasIntegralCorners B ∧ s ≤ B := by
+  have := Fintype.ofFinite ι
+  obtain ⟨R, hR₁, hR₂⟩ := IsBounded.subset_ball_lt h 0 0
+  let C : ℕ+ := ⟨⌈R⌉₊, Nat.ceil_pos.mpr hR₁⟩
+  let I : Box ι := BoxIntegral.Box.mk (fun _ ↦ - C) (fun _ ↦ C ) (fun _ ↦ by norm_num [hR₁])
+  refine ⟨I, ⟨fun _ ↦ - C, fun _ ↦ C, by aesop⟩, le_trans hR₂ ?_⟩
+  suffices Metric.ball (0 : ι → ℝ) C ≤ I from
+    le_trans (Metric.ball_subset_ball (Nat.le_ceil R)) this
+  intro x hx
+  simp_rw [mem_ball_zero_iff, pi_norm_lt_iff (by aesop : (0:ℝ) < C), Real.norm_eq_abs, abs_lt] at hx
+  exact fun i ↦ ⟨(hx i).1, le_of_lt (hx i).2⟩
+
+end hasIntegralCorners
+
+namespace BoxIntegral.unitPartition
+
+open Bornology MeasureTheory Fintype BoxIntegral
+
+variable (n : ℕ+)
+
+/-- A `BoxIntegral`, indexed by a positive integer `n` and `ν : ι → ℤ`, with corners `ν i / n`
+and of side length `1 / n`. -/
+def box (ν : ι → ℤ) : Box ι where
+  lower := fun i ↦ ν i / n
+  upper := fun i ↦ ν i / n + 1 / n
+  lower_lt_upper := fun _ ↦ by norm_num
+
+@[simp]
+theorem box_lower (ν : ι → ℤ) :
+    (box n ν).lower = fun i ↦ (ν i / n : ℝ) := rfl
+
+@[simp]
+theorem box_upper (ν : ι → ℤ) :
+    (box n ν).upper = fun i ↦ (ν i / n + 1 / n : ℝ) := rfl
+
+variable {n} in
+@[simp]
+theorem mem_box_iff {ν : ι → ℤ} {x : ι → ℝ} :
+    x ∈ box n ν ↔ ∀ i, ν i / n < x i ∧ x i ≤ ν i / n + 1 / n := by
+  simp_rw [Box.mem_def, box, Set.mem_Ioc]
+
+variable {n} in
+theorem mem_box_iff' {ν : ι → ℤ} {x : ι → ℝ} :
+    x ∈ box n ν ↔ ∀ i, ν i < n * x i ∧ n * x i ≤ ν i + 1 := by
+  have h_npos : 0 < (n:ℝ) := Nat.cast_pos.mpr <| PNat.pos n
+  simp_rw [mem_box_iff, ← _root_.le_div_iff₀' h_npos, ← div_lt_iff₀' h_npos, add_div]
+
+/-- The tag of a `unitPartition.Box`. -/
+abbrev tag (ν : ι → ℤ) : ι → ℝ := fun i ↦ (ν i + 1) / n
+
+@[simp]
+theorem tag_apply (ν : ι → ℤ) (i : ι) : tag n ν i = (ν i + 1) / n := rfl
+
+theorem tag_injective : Function.Injective (fun ν : ι → ℤ ↦ tag n ν) := by
+  refine fun _ _ h ↦ funext_iff.mpr fun i ↦ ?_
+  have := congr_arg (fun x ↦ x i) h
+  field_simp at this
+  exact this
+
+theorem tag_mem (ν : ι → ℤ) :
+    tag n ν ∈ box n ν := by
+  refine mem_box_iff.mpr fun _ ↦ ?_
+  rw [tag, add_div]
+  exact ⟨by norm_num, le_rfl⟩
```

Comment:
```suggestion
  exact ⟨lt_add_of_pos_right _ <| by positivity, le_rfl⟩
```
This is not really a job for `norm_num` (there are no computations with numerals involved).